controllers/cloud.redhat.com/errors/errors.go

```diff
@@ -73,22 +73,38 @@ func Wrap(msg string, err error) *ClowderError {
 	return clowderErr
 }
 
-// MissingDependencies is a struct containing a map of string lists, it is intended to hold
-// information about missing dependencies. The key is a resource type, or a provider as a string,
-// and the list is a list of keys or items that are missing.
+//MissingDependency is a struct that holds information about a missing dependency
+type MissingDependency struct {
+	Source  string
+	App     string
+	Details string
+}
+
+//ToString returns a string representation of the missing dependency
+func (m *MissingDependency) ToString() string {
+	return fmt.Sprintf("source: %s, app: %s, details: %s", m.Source, m.App, m.Details)
+}
+
+func MakeMissingDependencies(missingDep MissingDependency) MissingDependencies {
+	return MissingDependencies{
+		MissingDeps: []MissingDependency{missingDep},
+	}
+}
+
+//MissingDependencies is a struct that holds a list of MissingDependency structs
 type MissingDependencies struct {
-	MissingDeps map[string][]string
+	MissingDeps []MissingDependency
```

Comment:
I am pretty sure that this was a problem before, but as we are refactoring this, the missing deps can show up in a status field. Since this is a list the order should be determinate, but should we throw a sort in here to ensure that we don't get a string that can be different from the last time, such operations reduce the amount of reconciliations. I think there are potentially other places in the subsequent functions where we are picking up items from a loop. If those loops are looping over things which are not guaranteed to arrive in order, we may need to sort in there also to avoid creating reordered error lists.

Reply:
I'm afraid I don't understand this input. What would I sort by? Would I add a timestamp to `MissingDepdency` and then when we iterate through the `MissingDependencies.MissingDeps` (in the `MissingDepdencies.Error()` method) sort by time before creating the error strings? What does this solve? I thought this code just produced log messages? 

That said, this got me thinking about the `MissingDepdencies` object itself and I'm not sure why it contains a list at all. When I was refactoring this code I wasn't thinking about the larger picture of like how these structs are used, I was just trying to take what was there and improve it while changing as little as possible. But, as I go through the code I see no place where we ever add another `MissingDependency` to `MissingDepdencies.MissingDeps`. Every single place we create a `MissingDepdencies` instance we just create it with one `MissingDependency` - and we never add anything to it. Heck, there isn't even a method for adding a dependency after instantiation, and you can bet I would have added a method rather than allow access to raw object state.

So, while I don't fully understand your input I wonder if we should just scrap `MissingDependencies` all together and just work with `MissingDependency` instances, and add the `Error` method to `MissingDepdency` and then refactor `errors.HandleError` to expect the `MissingDepdency` type instead of `MissingDependencies` - I wonder if maybe this system was intended to handle cases where a single `MissingDependencies` instance would have multiple `MissingDependency` instances popped into it, but looking at the code it doesn't appear this ever actually came true. If this isn't a real world use case we should remove it and simplify.

Reply:
I can be too long winded: TL;DR - maybe we get rid of `MissingDependencies` all together and just pass `MissingDepdency` around because that list only ever contains one entry.

Reply:
Yeh I can see that, I think we kept it as a list in case we wanted to ever do a concurrent reconcile. - I'm happy to rename it. Now I remember the context, it was actually what was happening inside of the error itself. When we were trying to report kafka topic errors, the topics had to be sorted because otherwise each time we wrote the status, it changed and was different, causing needless writes to the etc database

Reply:
Do as you will with the rename

Reply:
I was wrong:

`providers/depdencies/impl.go`
```golang
	missingDeps := makeDepConfig(
		&depConfig,
		&privDepConfig,
		dep.Provider.Env.Spec.Providers.Web.Port,
		dep.Provider.Env.Spec.Providers.Web.PrivatePort,
		app,
		apps,
	)
	if len(missingDeps) > 0 {
		missingDepStructs := []errors.MissingDependency{}
		for _, dep := range missingDeps {
			missingDepStructs = append(missingDepStructs, errors.MissingDependency{
				Source:  "service",
				Details: dep,
			})
		}
		return &errors.MissingDependencies{MissingDeps: missingDepStructs}
	}
```

So there is one place where we use an missing deps with a list of missing deps; and all you need is one to justify leaving the API consistent. So, we should keep with what we have.

So I rescind my response about removing the list feature. So that brings me back to your input. I don't know how you'd want these sorted or in what way. It looks like `providers/depdencies/impl.go` would be where we'd do it. Maybe in `makeDepConfig` - let me know if this is still something you want for this patch and if so how to do it.

Reply:
I mean - it does now :) - that change is from your PR :)
Actually looking at it - in this case - we are looping over a static list of deps that comes from the clowdapp spec so it's not likely to change order. The problem will be if we bring in a new provider or something that sends back missing deps. If we do that we may want to put in a new object that has an `add` method, that will auto sort it after adding.